scripts/deploy.sh

```diff
@@ -0,0 +1,136 @@
+#! /usr/bin/env bash
+
+set -e
+
+message() {
+
+    echo
+    echo -----------------------------------------------------------------------------
+    echo "$@"
+    echo -----------------------------------------------------------------------------
+    echo
+}
+
+addValuesToFile() {
+    result=$(jq -s add "$1" /dev/stdin)
+    printf %s "$result" > "$1"
+}
+
+deploy() {
+    echo -e "Deploying $1\n"
+    DEPLOYED_ADDR=$(dapp create "$2" "${@:3}")
+    verify "$2" "$DEPLOYED_ADDR" "${@:3}"
+    echo -e "\nDeployed $1 to $DEPLOYED_ADDR\n"
+}
+
+verify() {
+  if [ -n "$ETHERSCAN_API_KEY" ]
+  then
+      sleep 5 # give etherscan some time to process the block
+      dapp verify-contract --async "$@"
+  else
+      echo "ETHERSCAN_API_KEY not provided, skipping verification"
+  fi
+}
+
+# Set up the defaults
+DEPLOYMENT_JSON=${DEPLOYMENT_JSON:-./deployment_$(seth chain).json}
+RESERVE_OWNER=${RESERVE_OWNER:-$ETH_FROM}
+DRIPS_HUB_ADMIN=${DRIPS_HUB_ADMIN:-$ETH_FROM}
+CYCLE_SECS=${CYCLE_SECS:-$(( 7 * 24 * 60 * 60 ))} # 1 week
+
+# Print the configuration
+message Deployment Config
+echo "Network:                  $(seth chain)"
+echo "Deployer address:         $ETH_FROM"
+echo "Gas price:                ${ETH_GAS_PRICE:-use the default}"
+ETHERSCAN_API_KEY_PROVIDED=${ETHERSCAN_API_KEY:+provided}
+echo "Etherscan API key:        ${ETHERSCAN_API_KEY_PROVIDED:-not provided}"
+echo "Deployment JSON:          $DEPLOYMENT_JSON"
+TO_DEPLOY="to be deployed"
+echo "Reserve:                  ${RESERVE:-$TO_DEPLOY}"
+echo "Reserve owner:            $RESERVE_OWNER"
+echo "DripsHub:                 ${DRIPS_HUB:-$TO_DEPLOY}"
+echo "DripsHub admin:           $DRIPS_HUB_ADMIN"
+echo "DripsHub logic:           ${DRIPS_HUB_LOGIC:-$TO_DEPLOY}"
+echo "DripsHub cycle seconds:   $CYCLE_SECS"
+echo "AddressApp:               ${ADDRESS_APP:-$TO_DEPLOY}"
+echo
+
+read -p "Ready to deploy? [y/n] " -n 1 -r
+echo
+if [[ $REPLY =~ ^[^Yy] ]]
+then
+    exit 1
+fi
+
+# Build the contracts
+message Building Contracts
+dapp build
+
+# Deploy the contracts
+message Deploying contracts
+
+if [ -z "$RESERVE" ]; then
+    deploy "Reserve" 'src/Reserve.sol:Reserve' "$ETH_FROM"
+    RESERVE=$DEPLOYED_ADDR
```

Comment:
Is it not possible to return a string from a bash function in an elegant way?

Reply:
The short answer is no, bash is a really terrible **programming** language :laughing: One often workaround is to `echo` the returned value in the function and then `$()` it on the caller side, this is what `drips-periphery` version does. Unfortunately it makes printing to the console from the inside of the function a little tricky, it can't use stdout anymore and it and all it can do is (ab)use stderr. I find a magic variable a little easier to control. It's also not unheard of in the bash world, many function do that, e.g. the built-in `read`.

Reply:
Ah, I see. Thanks for sharing.
I think it is okay. The script is not so complex and one can spot the pattern.